### PR TITLE
syncoid hold

### PR DESCRIPTION
**Pull request by rbike**
Added hold feature to syncoid.
Parameter "--use-hold" sets hold for newsyncsmap and removes hold from matchingsnap both on source and target. Hold name is "syncoid" + identitier + host name.
Setting holds on newsyncsnap prevents deletion of the of this snap until the next replication. In case of multiple targets a hold will be set for each target if different identifiers are used for each target. This prevents deletion until a new replications has been performed for all targets.

### CODE REVIEW

**Comment by jimsalterjrs** [comment]
Does this look good to you, @phreaker0 ? I haven't operationally tested it yet, but looks good at first blush.

**Comment by jimsalterjrs** [comment]
@rbike this will need updates to the in-program docs (which you see if you type `syncoid` with no arguments) as well, please.

**Comment by rbike** [comment]
> @rbike this will need updates to the in-program docs (which you see if you type `syncoid` with no arguments) as well, please.

@jimsalterjrs  I updated the in-program docs and readme.md with the --use-hold argument description. I also added code that checks that matchingsnap exists before removing the hold. In case of an initial replication a matchingsnap doesn't exist.

File: syncoid [hunk 1]
```diff
@@ -3,6 +3,9 @@
 # this software is licensed for use under the Free Software Foundation's GPL v3.0 license, as retrieved
 # from http://www.gnu.org/licenses/gpl-3.0.html on 2014-11-17.  A copy should also be available in this
 # project's Git repository at https://github.com/jimsalterjrs/sanoid/blob/master/LICENSE.
+# 
+# 2021-03-03: redbike
+# support for zfs holds added
```
Comment (phreaker0) [inline syncoid:8]: please remove this here, you will be credited in the changelog after the next release which includes the patch

File: syncoid [hunk 2]
```diff
@@ -827,7 +831,42 @@ sub syncdataset {
 			#setzfsvalue($targethost,$targetfs,$targetisroot,'readonly',$originaltargetreadonly);
 		}
 	}
-
+# if "--use-hold" parameter is used set hold on newsync snapshot and remove hold on matching snapshot both on source and target
```
Comment (phreaker0) [inline syncoid:834]: indentation needed here

File: syncoid [hunk 3]
```diff
@@ -1325,7 +1364,7 @@ sub buildsynccmd {
 sub pruneoldsyncsnaps {
 	my ($rhost,$fs,$newsyncsnap,$isroot,@snaps) = @_;
 
-	my $fsescaped = escapeshellparam($fs);
+	my $fsescaped = escapeshellparam($fs);
```
Comment (phreaker0) [inline syncoid:1367]: remove the added whitespace

File: syncoid [hunk 4]
```diff
@@ -1972,6 +2011,7 @@ Options:
   --no-sync-snap        Does not create new snapshot, only transfers existing
   --keep-sync-snap      Don't destroy created sync snapshots
   --create-bookmark     Creates a zfs bookmark for the newest snapshot on the source after replication succeeds (only works with --no-sync-snap)
+  --use-hold            Adds a hold to the newest snapshot on the source and target after replication succeeds and removes the hold after the next succesful replication. The hold name incldues the identifier if set. This allows for separate holds in case of multiple targets
```
Comment (phreaker0) [inline syncoid:2014]: please also document in the README.md

**Review by phreaker0 — CHANGES_REQUESTED** [review]
I played around with it and it works fine, just some minor cosmetic stuff needs to be adapted as commented.

File: syncoid (hunk 4 above)
Comment (rbike) [reply syncoid:2014]: @phreaker0 an updated readme.md is already provided

Comment (phreaker0) [reply syncoid:2014]: sry!, I somehow missed that

**Comment by rbike** [comment]
Implemented all changes proposed by @phreaker0. Should be ready for merging.

**Comment by rbike** [comment]
Hi @phreaker0 and @jimsalterjrs , is still some to be doner is it ready for merge?

**Review by phreaker0 — APPROVED** [review]
@rbike i rebased your PR to the latest master, thanks for the adaptions. Looks ready to me

**Pull request merged**